homeassistant/components/repetier/__init__.py

```diff
@@ -0,0 +1,240 @@
+        'state': {
+            'job': 'job_name',
+            'start': 'start',
+            'printtime': 'print_time',
+            'printedtimecomp': 'from_start',
+        },
+    },
+    'job_start': {
+        'offline': {
+            'job': None,
+            'state': 'off',
+            'start': None,
+            'printedtimecomp': None
+        },
+        'state': {
+            'job': 'job_name',
+            'start': 'start',
+            'printedtimecomp': 'from_start',
+        },
+    },
+}
+
+
+def has_all_unique_names(value):
+    """Validate that printers have an unique name."""
+    names = [util_slugify(printer[CONF_NAME]) for printer in value]
+    vol.Schema(vol.Unique())(names)
+    return value
+
+
+SENSOR_TYPES = {
+    # Type, Unit, Icon
+    'bed_temperature': ['temperature', TEMP_CELSIUS, 'mdi:thermometer',
+                        '_bed_'],
+    'extruder_temperature': ['temperature', TEMP_CELSIUS, 'mdi:thermometer',
+                             '_extruder_'],
+    'chamber_temperature': ['temperature', TEMP_CELSIUS, 'mdi:thermometer',
+                            '_chamber_'],
+    'current_state': ['state', None, 'mdi:printer-3d', ''],
+    'current_job': ['progress', '%', 'mdi:file-percent', '_current_job'],
+    'job_end': ['progress', None, 'mdi:clock-end', '_job_end'],
+    'job_start': ['progress', None, 'mdi:clock-start', '_job_start'],
+}
+
+SENSOR_SCHEMA = vol.Schema({
+    vol.Optional(CONF_MONITORED_CONDITIONS, default=list(SENSOR_TYPES)):
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=3344): cv.port,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_SENSORS, default={}): SENSOR_SCHEMA,
+    })], has_all_unique_names),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the Repetier Server component."""
+    import pyrepetier
+
+    hass.data[REPETIER_API] = {}
+    sensor_info = []
+
+    for repetier in config[DOMAIN]:
+        _LOGGER.debug("Repetier server config %s", repetier[CONF_HOST])
+
+        url = "http://{}".format(repetier[CONF_HOST])
+        port = repetier[CONF_PORT]
+        api_key = repetier[CONF_API_KEY]
+
+        client = pyrepetier.Repetier(
+            url=url,
+            port=port,
+            apikey=api_key)
+
+        printers = client.getprinters()
+
+        if not printers:
+            return False
+
+        api = PrinterAPI(hass, client, printers)
+        api.update()
+        track_time_interval(hass, api.update, SCAN_INTERVAL)
+
+        hass.data[REPETIER_API][repetier[CONF_NAME]] = api
+
+        sensors = repetier[CONF_SENSORS][CONF_MONITORED_CONDITIONS]
+        for pidx, printer in enumerate(printers):
```

Comment:
There might be an opportunity to break out this part into a helper method in the `PrinterAPI` class, and call this helper from `update` also in the same class. I would try that.

It should be able to select between new and existing printer sensors and either call `load_platform` or `dispatcher_send`.

Reply:
I have tried coding my version of this - see if it fits your idea